core/rawdb/accessors_chain.go

```diff
@@ -1749,3 +1820,84 @@ func DeleteUtxoToBlockHeight(db ethdb.KeyValueWriter, txHash common.Hash, index
 		db.Logger().WithField("err", err).Fatal("Failed to delete utxo to block height")
 	}
 }
+
+func ReadCoinbaseLockup(db ethdb.KeyValueReader, batch ethdb.Batch, ownerContract common.Address, beneficiaryMiner common.Address, lockupByte byte, epoch uint32) (*big.Int, uint32, uint16, common.Address) {
+	deleted, data := batch.GetPending(CoinbaseLockupKey(ownerContract, beneficiaryMiner, lockupByte, epoch))
+	if deleted {
+		return new(big.Int), 0, 0, common.Zero
+	} else if data != nil {
+		amount := new(big.Int).SetBytes(data[:32])
+		blockHeight := binary.BigEndian.Uint32(data[32:36])
+		elements := binary.BigEndian.Uint16(data[36:38])
+		var delegate common.Address
+		if len(data) == 58 {
+			delegate = common.BytesToAddress(data[38:], db.Location())
+		} else {
+			delegate = common.Zero
+		}
+		return amount, blockHeight, elements, delegate
+	}
+	// If the data is not in the batch, try to look up the data in leveldb
+	data, _ = db.Get(CoinbaseLockupKey(ownerContract, beneficiaryMiner, lockupByte, epoch))
+	if len(data) == 0 {
+		return new(big.Int), 0, 0, common.Zero
+	}
+	amount := new(big.Int).SetBytes(data[:32])
+	blockHeight := binary.BigEndian.Uint32(data[32:36])
+	elements := binary.BigEndian.Uint16(data[36:38])
+	var delegate common.Address
+	if len(data) == 58 {
+		delegate = common.BytesToAddress(data[38:], db.Location())
+	} else {
+		delegate = common.Zero
+	}
+	return amount, blockHeight, elements, delegate
+}
+
+func WriteCoinbaseLockup(db ethdb.KeyValueWriter, ownerContract common.Address, beneficiaryMiner common.Address, lockupByte byte, epoch uint32, amount *big.Int, blockHeight uint32, elements uint16, delegate common.Address) ([]byte, error) {
+	data := make([]byte, 38)
+	amountBytes := amount.Bytes()
+	if len(amountBytes) > 32 {
```

Comment:
is this the correct check? its value in wei right? that means 2^32 is actually a small value in quai. 4.294967296 × 10^-9 wolfram says